Adding outbound flow control for Netty.

nmittler opened the pull request.

nmittler commented:
@louiscryan @ejona86 can you take a look?  I still have to add tests for outbound flow control, but I wanted to get thoughts on the implementation first.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -42,7 +44,13 @@
+   * The default number of queued bytes for a given stream, below which
+   * {@link StreamListener#onReady(int)} will be called.
+   */
+  public static final int DEFAULT_ONREADY_THRESHOLD = 1048576;
```
This can be a much lower number, since the bandwidth-delay product between the application and the transport is the _network's_ bandwidth * context switch latency.

Netty [uses 32K for instance](https://github.com/netty/netty/blob/8f8a06ab0a885accf72b19cb27f214072204726b/transport/src/main/java/io/netty/channel/DefaultChannelConfig.java#L62).

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
+      // TODO(nathanmittler): Consider selecting a better value for numMessages.
+      listener().onReady(1);
```
I tried to communicate this earlier, but this doesn't work when an application exceeds the number of tokens provided.

1. We call `onReady(1)` (token total: 1)
2. Application sends two messages, each greater than `onReadyThreshold` (token total: -1)
3. After sending, eventually we call `onReady(1)` (token total: 0)
4. Deadlock, as application is waiting for another token.

There are two solutions:

1. Get rid of tokens. Just have zero-arg `onReady()`. If the application hasn't sent after the last onReady(), they can send more. If they sent more than once consecutively, then the application will over-buffer, but everything continues functioning.
2. Count how many messages are in each frame, and make sure to return that number of tokens.

Netty's outbound flow control is like option 1 (and so that is what my "streams as Channels" PR would provide... eventually).

nmittler replied:
I think option 1 is probably the most reasonable here. Counting the number of messages in each frame will get tricky especially when compressing.  WDYT?

ejona86 replied:
I agree; option 1 seems to be the preferred way to go.

Note that compression doesn't really add any complexity; you can have issues just with small messages.

nmittler replied:
Done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -150,6 +186,11 @@ public void dispose() {
   /**
+   * Gets the listener to this stream.
+   */
+  protected abstract LT listener();
```
Just make this return StreamListener? It doesn't seem we need it to be generic at all, or did I miss something?

nmittler replied:
Done.  I think this was an artifact of an earlier version.

nmittler commented:
@ejona86 PTAL

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+   */
+  protected final void onSentBytes(int numBytes) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
```
This will end up notifying for each frame we sent after the queue is less than onReadyThreshold. We probably only want to notify if there has been another write (`onSendingBytes()`)?

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    numSentBytesQueued += numBytes;
```
We should probably call `onReady()` here if `numSentBytesQueued < onReadyThreshold` ? We don't want to wait until the write actually completes.

nmittler replied:
Good call. done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+  protected final void onSentBytes(int numBytes) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
+      // TODO(nathanmittler): Consider selecting a better value for numMessages.
```
This comment seems like it can be removed.

nmittler replied:
Already did :) ... refresh.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -104,6 +123,23 @@ public void deliverFrame(WritableBuffer frame, boolean endOfStream, boolean flus
+   */
+  public void setOnReadyThreshold(int onReadyThreshold) {
+    checkArgument(onReadyThreshold > 0, "onReadyThreshold must be > 0");
+    this.onReadyThreshold = onReadyThreshold;
```
This needs to check whether onReady() needs to be called, or it needs to have restrictions on when it can be called.

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
+      // TODO(nathanmittler): Consider selecting a better value for numMessages.
+      listener().onReady();
```
If CLOSED, then we must not call `onReady()`.

On client-side, it isn't strictly necessary to call `onReady()` if HALF_CLOSED, but that seems like it would be a bit of a pain, since this is shared between client and server. I'm okay if we don't do this for now.

ejona86 commented on netty/src/main/java/io/grpc/transport/netty/NettyServerStream.java:
```diff
@@ -90,7 +92,24 @@ protected void internalSendHeaders(Metadata.Headers headers) {
     ByteBuf bytebuf = ((NettyWritableBuffer) frame).bytebuf();
-    channel.write(new SendGrpcFrameCommand(this, bytebuf, endOfStream));
+    final int numBytes = bytebuf.readableBytes();
+    if (numBytes > 0) {
```
I'd prefer not to have a `numBytes == 0` optimization. It shouldn't ever happen on server-side right now, so the else won't ever be tested/run in practice.

nmittler replied:
done.

nmittler commented:
@ejona86 PTAL

ejona86 commented on netty/src/main/java/io/grpc/transport/netty/NettyClientStream.java:
```diff
@@ -117,13 +122,25 @@ protected void sendCancel() {
+    final int numBytes = bytebuf.readableBytes();
+    if (numBytes > 0) {
+      // Add the bytes to outbound flow control.
+      onSendingBytes(numBytes);
```
Hmm... This isn't thread-safe, as we aren't in the event loop here.

nmittler replied:
Right, good catch.  Added synchronization.

nmittler commented:
@ejona86 PTAL, I added synchronization.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+   */
+  protected final void onSentBytes(int numBytes) {
+    boolean shouldNotify = false;
+    synchronized(onReadyLock) {
```
Need space before (

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+   * @param numBytes the number of bytes that were sent.
+   */
+  protected final void onSentBytes(int numBytes) {
+    boolean shouldNotify = false;
```
Could we not initialize this variable here, since its value will always be overwritten?

nmittler replied:
done.

nmittler commented:
@ejona86 changes committed ... anything else at this point?  If not, I'll start adding unit tests.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+      shouldNotify = shouldNotifyOnReady();
+    }
+    if (shouldNotify) {
+      listener().onReady();
```
Hmm... We can't actually call listener().onReady() here, as we are only supposed to call the listener from the transport thread. Nonetheless, it seems it may be safe from a threading perspective, but we would need to update documentation. There's also the possibility of recursive loops (application sends -> onReady() -> application sends -> onReady()) and application code may not be reentrant.

Would it be better to call onSendingBytes() within the event loop? However, it is nice to be able to know during writing that the application could continue sending, which would reduce latency.

Or maybe we should return a boolean for whether the buffer is full? We would only call onReady() if we returned that the buffer filled. If we had a method isReady(), that would be the Netty behavior, but I'm not sure it applies as well for us.

:-/ all around.

nmittler replied:
For efficiency of applications that obey outbound flow control, I think that immediately calling back during a write (and requiring the application code to be reentrant) is probably the best approach, so long as the documentation is clear. WDYT?

ejona86 replied:
My feeling is that would lead to a lot of recursion, such that it could exhaust the stack. At 32K, we can fit 6.5K messages in the buffer.

If we are doing it for efficiency, then I think that returning a boolean would be a better approach. Doing a callback requires multiple allocations, has the reentrancy/recursion problem, has more synchronization in grpc, and needs more application synchronization between sender and listener. From a threading perspective, only doing callbacks from transport thread would also be a plus.

nmittler commented:
@ejona86 I've rebased my changes and changed the contract so that `onReady` is always called from the transport thread.  I've also changed `Stream.writeMessage` to return a `boolean` to enable the immediate notification case.

/cc @louiscryan FYI

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -113,10 +159,14 @@ public void writeMessage(InputStream message, int length, @Nullable Runnable acc
+    // Indicate whether or not the transport is ready to send additional data. Calling this
+    // method has the side-effect of unsetting the shouldNotifyOnReady flag, which is fine since
+    // returning true here is equivalent to calling onReady anyway.
+    return shouldNotifyOnReady();
```
This won't work like you expect, since the framer will basically never have written out the frame at this point. Instead, I guess flush() should return the boolean.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -113,10 +159,14 @@ public void writeMessage(InputStream message, int length, @Nullable Runnable acc
       framer.writePayload(message, length);
```
So if the stream is closed, we will end up returning false. We may want that to be true to drain any application buffers, but it could go either way. We would want NoopClientStream to match though.

ejona86 commented:
@nmittler, this approach looks like it will work.

Doing the warning cleanup in a separate PR is important though. It is really making this PR noisy, which makes it much harder to review.

nmittler commented:
PTAL ... I've rebased the changes now that the warning fixes are in.

ejona86 commented on core/src/main/java/io/grpc/ClientInterceptors.java:
```diff
@@ -141,7 +141,9 @@ public void cancel() { }
     @Override
-    public void sendPayload(Object payload) { }
+    public boolean sendPayload(Object payload) {
+      return true;
```
We should do the same thing here and for closed calls. Right now though one returns false and the other returns true.

nmittler replied:
@ejona86 I'm not seeing what you mean?  Which one returns false?

nmittler replied:
oh you mean ServerCall ... yeah, let me add this change there as well.

ejona86 replied:
isReady() returns false if closed.

ejona86 replied:
I wasn't meaning ServerCall, that was separate.

nmittler replied:
Done.

nmittler replied:
@ejona86 what do you suggest we do here? It maintains no state.

ejona86 replied:
Treat it the same as closed. If one returns true, then the other should return true. If false, false.

nmittler replied:
I'm not getting what you're saying ... there is no `closed` flag that I can look at here. :/

ejona86 replied:
Right now `isReady()` returns `false` if the stream is closed. Thus, I'm saying that this should be `return false;`. There is no need for a condition.

If `return false;` seems wrong, then `isReady()` should be changed to return `true` if the stream is closed.

nmittler replied:
Oh ok ... the detail you were leaving out (and the source of my confusion) was that this is only used when an exception occurs in the `start` method ... meaning the call is closed.  So in this case it makes sense to always return `false`.

nmittler replied:
Done.

ejona86 commented:
Change ServerCall.sendPayload() as well?

nmittler commented:
@ejona86 committed updates for ServerCall ... would be nice if client/server shared a common interface hierarchy :)

nicolasnoble commented:
Can one of the admins verify this patch?

nmittler commented:
@ejona86 PTAL ... if you're happy with this, I'll begin unit testing.

ejona86 commented:
@nmittler, yeah, things look good; I'm happy with it.

nmittler commented:
@ejona86 PTAL.  Added unit tests and made the Call/ServerCall interfaces consistent with the transport (i.e. `isReady()`). Also changed the behavior of `isReady` so that it is idempotent, which I think is what would be expected.

ejona86 commented on core/src/main/java/io/grpc/transport/AbstractStream.java:
```diff
@@ -229,6 +288,68 @@ protected final void requestMessagesFromDeframer(int numMessages) {
+   */
+  protected final void onSendingBytes(int numBytes) {
+    synchronized (onReadyLock) {
+      shouldNotifyOnReady = true;
```
If you are going to have a separate isReady() call, then we only want to call onReady() if isReady() could have returned false.

I'd suggest this:
```
numSentBytesQueued += numBytes;
if (numSentBytesQueued >= onReadyThreshold) {
  shouldNotifyOnReady = true;
}
```

You could also use:
```
numSentBytesQueued += numBytes;
if (!isReady()) {
  shouldNotifyOnReady = true;
}
```

nmittler replied:
Done.

nmittler commented:
@ejona86 PTAL

ejona86 commented:
@nmittler, LGTM! Woo!

nmittler commented:
cherry-picked as 64176d5560a1041db9495f4b3a3c6e54aaf43ea8
